handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -836,98 +869,76 @@ public final SSLEngineResult unwrap(
-                            // we will produce an error on the next write attempt, which will trigger a SSL.clearError()
-                            // later.
-                            break;
+                            if (packetLength != 0) {
+                                // Not completely written into the BIO yet continue writing.
+                                continue;
+                            }
                         }
-                    } else {
-                        // BIO_write returned a negative or zero number, this means we could not complete the write
-                        // operation and should retry later.
-                        // We ignore BIO_* errors here as we use in memory BIO anyway and will do another SSL_* call
-                        // later on in which we will produce an exception in case of an error
-                        SSL.clearError();
-                        break;
                     }
-                } while (srcsOffset < srcsEndOffset);
-                bytesConsumed = packetLength - packetLengthRemaining;
-            }
-
-            // Number of produced bytes
-            int bytesProduced = 0;
 
-            if (capacity > 0) {
-                // Write decrypted data to dsts buffers
-                int idx = dstsOffset;
-                while (idx < endOffset) {
-                    ByteBuffer dst = dsts[idx];
-                    if (!dst.hasRemaining()) {
-                        idx++;
-                        continue;
-                    }
+                    // Write decrypted data to dsts buffers
+                    readLoop:
+                    while (dstsOffset < endOffset) {
+                        ByteBuffer dst = dsts[dstsOffset];
+                        if (!dst.hasRemaining()) {
+                            // No space left in the destination buffer, skip it.
+                            dstsOffset++;
+                            continue;
+                        }
 
-                    int bytesRead = readPlaintextData(dst);
+                        int bytesRead = readPlaintextData(dst);
 
-                    // TODO: We may want to consider if we move this check and only do it in a less often called place
-                    // at the price of not being 100% accurate, like for example when calling SSL.getError(...).
-                    rejectRemoteInitiatedRenegation();
+                        if (bytesRead > 0) {
+                            bytesProduced += bytesRead;
 
-                    if (bytesRead > 0) {
-                        bytesProduced += bytesRead;
+                            if (!dst.hasRemaining()) {
+                                // Move to the next dst buffer as this one is full.
+                                dstsOffset++;
+                            } else {
+                                if (packetLength == 0) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS134)